sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,273 @@
+        strictly higher than the given threshold. If float, the parameter
+        represents a proportion of documents, integer represents absolute
+        counts.
+
+    Attributes
+    ----------
+    file_names_ : list of str
+        The list of file names used for computing the similarity matrix.
+        The recommended examples are chosen among this list.
+
+    similarity_matrix_ : dense matrix
+        Fitted matrix of pairwise cosine similarities.
+    """
+
+    def __init__(self, *, n_examples=5, min_df=3, max_df=0.9):
+        self.n_examples = n_examples
+        self.min_df = min_df
+        self.max_df = max_df
+
+    def token_freqs(self, doc):
+        """Extract a dict mapping raw tokens from doc to their occurrences."""
+        token_generator = (tok.lower() for tok in re.findall(r"\w+", doc))
+        return self.dict_freqs(token_generator)
+
+    @staticmethod
+    def dict_freqs(doc):
+        """Extract a dict mapping list of tokens to their occurrences."""
+        freq = defaultdict(int)
+        for tok in doc:
+            freq[tok] += 1
+        return freq
+
+    def dict_vectorizer(self, data):
+        """Convert a dictionary of feature arrays into a matrix.
+
+        Parameters
+        ----------
+        data : list of dict
+            An iterable of dictionaries of feature arrays, where each key
+            corresponds to a feature name, and each value is an array of
+            feature values.
+
+        Returns
+        -------
+        X : ndarray of shape (n_samples, n_features)
+            A matrix of occurrences where n_samples is the number of samples in
+            the dataset and n_features is the total number of features across
+            all samples.
+        """
+        try:
+            import numpy as np
+
+        except ImportError:
+            raise ConfigError("gallery_conf['recommender'] requires numpy")
+
+        feature_names = []
+        all_values = defaultdict(list)
+        for row in data:
+            for feature_name, feature_value in row.items():
+                feature_names.append(feature_name)
+                all_values[feature_name].append(feature_value)
+
+        feature_dict = {feature: i for i, feature in enumerate(sorted(all_values))}
+        X = np.zeros((len(data), len(feature_dict)))
+        for idx, row in enumerate(data):
+            for feature_name, feature_value in row.items():
+                X[idx, feature_dict[feature_name]] = feature_value
+        return X
+
+    def compute_tf_idf(self, X):
+        """Transform a term frequency matrix into a TF-IDF matrix.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_features)
+            A term frequency matrix.
+
+        Returns
+        -------
+        X_tfidf : ndarray of shape (n_samples, n_features)
+            A tf-idf matrix of the same shape as X.
+        """
+        import numpy as np
+
+        n_samples, n_features = X.shape
```

Comment:
Nitpick, we don't use `n_features`:

```suggestion
        n_samples = X.shape[0] + 1
```
we can remove line 127, and maybe add a comment that `+1` is to avoid taking log of 0.